### PR TITLE
Add picker source

### PR DESCRIPTION
**Pull request by syoung-smallwisdom**
Added support for using a picker to set input values for numbers, dates, multiple component, choices, weight, and height.

Additionally, I discovered during testing that unit conversions weren't working correctly and have corrected that for certain types (namely, height and weight).

Note: If the user enters a height/weight that has a non-English unit symbol, that unit is currently ignored, but entering "60 kg" even where the formatter is setup to expect "136 lb" will still work. Baby steps. ;)

### CODE REVIEW

File: ResearchSuite/ResearchSuite/RSDLengthFormatter.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+///
+/// - note: While this SDK is written in Swift where permissibile, formatters are written in Obj-c to allow
+/// overriding `-getObjectValue:forString:errorDescription:`. Apple documentation does not include how to
+/// set the value of the pointer for a Swift 4 implementation of the function. syoung 12/30/2017
+///
+@interface RSDLengthFormatter : NSLengthFormatter
+
+/// Is this formatter used to describe a child's height? The default behavior for `NSLengthFormatter`
+/// when the locale uses imperial units (not metric) is to show for a person's height in ft/in even
+/// if it is for an infant or child. Whereas, typically a US height for infants and children is shown
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDLengthFormatter.h]: ...child. Whereas, typically... --> ...child, whereas typically...
also maybe "small children" or "toddlers" (or just leave it at "infants")? Definitely by age 4 or 5 they're measured in feet/inches.

File: ResearchSuite/ResearchSuite/RSDLengthFormatter.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+/// the locale is metric, in which case it is ignored.
+///
+/// - note: Setting this value to `true` will also set `isForPersonHeightUse` to `true`.
+///
+@property (nonatomic, getter=isForChildHeightUse) BOOL forChildHeightUse;
+
+/// The base unit to use for converting a `NSNumber` to a string. The default value will depend upon
+/// whether or not this is used for measuring a person's height, and whether or not the `numberFormatter`
+/// locale uses the metric system. If this is a height formatter then the default metric unit is
+/// centimeters and the default imperial unit (US English) is inches. Otherwise, the default metric
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDLengthFormatter.h]: Digression for an esoteric discussion of unit systems: 

The US doesn't use Imperial units, but US Customary units. For length, and for weight units up to the pound, these are by international agreement identical, but for higher weight units (e.g. cwt.) and for volumes they are often not only wildly different between US and Imperial for units with the same name, but also sometimes within the US system (a US dry gallon, and thus a dry cup/ounce/tbsp/tsp, is a different volume from a US liquid gallon/cup/ounce/tbsp/tsp).

I don't think we've had any use cases for volume measurements yet (though they would be applicable if nutrition data were collected in a study).

tl;dr
For our current purposes, I think we're mostly using units where they are the same (the exception being that humans' weights in the UK are usually given in stone/lbs for adults and older children, and lbs/oz for infants) so maybe describe them as US/Imperial (or Imperial/US) units? Or maybe this is too pedantic and it could just be Imperial (capitalized). I'll leave that to you as being more objective since this is obviously a pet peeve of mine. ;)

File: ResearchSuite/ResearchSuite/RSDLengthFormatter.h [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+///
+/// This is used by the `-stringForObjectValue:` method to convert the input object to a localized string
+/// for the case where the object value is a `NSNumber`. If the input object is a `NSMeasurement` then
+/// this value is ignored.
+@property (nonatomic) NSUnitLength *toStringUnit;
+
+/// The base unit to use for converting a string to an `NSMeasurement`. The default value will depend upon
+/// whether or not this is used for measuring a person's height, and whether or not the `numberFormatter`
+/// locale uses the metric system. If this is a height formatter then the default metric unit is
+/// centimeters and the default imperial unit (US English) is inches. Otherwise, the default metric
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDLengthFormatter.h]: Unit system name (also on next line)

File: ResearchSuite/ResearchSuite/RSDMassFormatter.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/// `RSDMassFormatter` is a custom subclass of the `NSMassFormatter` that can convert a `NSMeasurement`
+/// object to a localized string.
+///
+/// The default converters are overriden to check for a `NSMeasurement` and return the converted string
+/// if available. If the units are imperial for the current locale, this is for measuring the height
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDMassFormatter.h]: Unit system name

File: ResearchSuite/ResearchSuite/RSDMassFormatter.h [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+/// - note: While this SDK is written in Swift where permissibile, formatters are written in Obj-c to allow
+/// overriding `-getObjectValue:forString:errorDescription:`. Apple documentation does not include how to
+/// set the value of the pointer for a Swift 4 implementation of the function. syoung 12/30/2017
+///
+@interface RSDMassFormatter : NSMassFormatter
+
+/// Is this formatter used to describe an infant's weight (mass)? The default behavior for
+/// `NSMassFormatter` when the locale uses imperial units (not metric) is to show for a person's weight
+/// in lb even if it is for an infant. Whereas, typically a US weight for infants  is shown in lb and oz.
+/// This property, if set to `true` will return weight in lb and oz unless the locale is metric,
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDMassFormatter.h]: `true` --> `true`,

File: ResearchSuite/ResearchSuite/RSDMassFormatter.h [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+///
+/// - note: While this SDK is written in Swift where permissibile, formatters are written in Obj-c to allow
+/// overriding `-getObjectValue:forString:errorDescription:`. Apple documentation does not include how to
+/// set the value of the pointer for a Swift 4 implementation of the function. syoung 12/30/2017
+///
+@interface RSDMassFormatter : NSMassFormatter
+
+/// Is this formatter used to describe an infant's weight (mass)? The default behavior for
+/// `NSMassFormatter` when the locale uses imperial units (not metric) is to show for a person's weight
+/// in lb even if it is for an infant. Whereas, typically a US weight for infants  is shown in lb and oz.
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDMassFormatter.h]: ...infant. Whereas, typically... --> infant, whereas typically

File: ResearchSuite/ResearchSuite/RSDMassFormatter.m [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        double pounds = floor([measurement measurementByConvertingToUnit:NSUnitMass.poundsMass].doubleValue);
+        double ounces = [measurement measurementByConvertingToUnit:NSUnitMass.ounces].doubleValue - (pounds * 16.0);
+        ounces = round(1000 * ounces) / 1000;
+        NSString * poundString = [self stringFromValue:pounds unit:NSMassFormatterUnitPound];
+        NSString * ouncesString = [self stringFromValue:ounces unit:NSMassFormatterUnitOunce];
+        NSString * joinedString = [NSString localizedStringWithFormat:@"%@, %@", poundString, ouncesString];
+        return joinedString;
+        
+    } else {
+        // Convert the value to a number and use the default impplementation from the parent class.
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDMassFormatter.m:98]: impplementation --> implementation

File: ResearchSuite/ResearchSuite/RSDMassFormatter.m [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+        self.fromStringUnit == castObject.fromStringUnit;
+    }
+}
+
+- (NSUInteger)hash {
+    return [super hash] ^ self.isForInfantMassUse ^ self.toStringUnit.hash ^ self.fromStringUnit.hash;
+}
+
+@end
+
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDMassFormatter.m:216]: extra newlines at end of file

File: ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift [hunk 9]
```diff
@@ -34,7 +34,7 @@
 import Foundation
 
 /// Extension of the `RSDMultipleComponentInputField` protocol to implement the `RSDChoicePickerDataSource` protocol.
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift]: Is this comment still applicable/in the right place?

File: ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift [hunk 10]
```diff
@@ -124,26 +193,211 @@ public struct RSDNumberPickerDataSourceObject : RSDNumberPickerDataSource {
...
+    /// Returns the text answer to display for a given selected answer.
+    /// - parameter selectedAnswer: The answer to convert.
+    /// - returns: A text value for the answer to display to the user.
+    public func textAnswer(from selectedAnswer: Any?) -> String? {
+        guard let number = numberAnswer(from: selectedAnswer) else { return nil }
+        return numberFormatter.string(from: number as NSNumber)
     }
+}
+
+/// `RSDImperialMeasurementPickerDataSource` is a generic protocol for converting imperial (multiple component)
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift]: prior comment regarding Imperial vs US Customary may or may not be relevant to the naming of this protocol...

File: ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift [hunk 11]
```diff
@@ -124,26 +193,211 @@ public struct RSDNumberPickerDataSourceObject : RSDNumberPickerDataSource {
...
+/// units from and to imperial units.
+public protocol RSDImperialMeasurementPickerDataSource : RSDMultipleComponentOptions {
+    associatedtype UnitType : Dimension
     
-    // Returns the choice for this row/component.
-    public func choice(forRow row: Int, forComponent component: Int) -> RSDChoice? {
-        fatalError("Not yet implemented")
+    // The imperial unit converter.
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift]: the imperial thing again

File: ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift [hunk 12]
```diff
@@ -124,26 +193,211 @@ public struct RSDNumberPickerDataSourceObject : RSDNumberPickerDataSource {
...
+                return nil
+        }
+        return [Int(tuple.largeValue) - largeUnitBounds.lower, Int(tuple.smallValue)]
+    }
+}
+
+/// `RSDUSHeightPickerDataSource` is a custom height picker for use when the `Locale` is `US_en`.
+public struct RSDUSHeightPickerDataSource : RSDImperialMeasurementPickerDataSource {
+    
+    // The imperial unit converter.
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift]: imperial --> ??

File: ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift [hunk 13]
```diff
@@ -124,26 +193,211 @@ public struct RSDNumberPickerDataSourceObject : RSDNumberPickerDataSource {
...
+    public let converter: RSDUnitConverter.ImperialConverter<UnitLength>
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift]: ditto

File: ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift [hunk 14]
```diff
@@ -124,26 +193,211 @@ public struct RSDNumberPickerDataSourceObject : RSDNumberPickerDataSource {
...
+            }
+        }
+    }
+}
+
+/// `RSDUSInfantMassPickerDataSource` is a custom weight picker for use when the `Locale` is `US_en`
+/// and the mass is for an infant in "lb, oz".
+public struct RSDUSInfantMassPickerDataSource : RSDImperialMeasurementPickerDataSource {
+    
+    // The imperial unit converter.
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift]: [play Star Wars Imperial March]

File: ResearchSuite/ResearchSuite/RSDTableItem.swift [hunk 15]
```diff
@@ -390,3 +404,162 @@ final class RSDNumberInputTableItem : RSDTextInputTableItem {
         return answer
     }
 }
+
+/// A custom implementation for inputing a person's mass (weight) that can set up a picker for
+/// setting the weight of an infant. Typically, US English user's know what their newborn baby's
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDTableItem.swift]: user's --> users (or better still, participants)

File: ResearchSuite/ResearchSuite/RSDTableItem.swift [hunk 16]
```diff
@@ -390,3 +404,162 @@ final class RSDNumberInputTableItem : RSDTextInputTableItem {
...
+            massUnit = massFormatter.toStringUnit
+        }
+        else {
+            massUnit = ((unit != nil) ? UnitMass(fromSymbol: unit!) : nil) ?? .kilograms
+        }
+        
+        self.baseUnit = massUnit
+        unit = massUnit.symbol
+        
+        // If the measurement is for an infant and the local is US
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDTableItem.swift]: locale

File: ResearchSuite/ResearchSuite/RSDTableItem.swift [hunk 17]
```diff
@@ -390,3 +404,162 @@ final class RSDNumberInputTableItem : RSDTextInputTableItem {
...
+    /// by the formatter into a decimal value in the `baseUnit`.
+    override func convertAnswer(_ newValue: Any) throws -> Any? {
+        let answer = try super.convertAnswer(newValue)
+        guard let measurement = answer as? Measurement<UnitMass> else { return answer }
+        return measurement.converted(to: baseUnit).value
+    }
+}
+
+/// A custom implementation for inputing a person's height that can set up a picker for
+/// setting the height of an adult. Typically, US English user's will report an adult height
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDTableItem.swift]: user's --> users (or participants)

File: ResearchSuite/ResearchSuite/RSDTableItem.swift [hunk 18]
```diff
@@ -390,3 +404,162 @@ final class RSDNumberInputTableItem : RSDTextInputTableItem {
...
+/// using ft and in, whereas a child or infant will be reported in inches. The placeholder
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDTableItem.swift]: small child (or toddler), or "under 4 ft" (or whatever the threshold is, but it's somewhere around there)

File: ResearchSuite/ResearchSuite/RSDUnitConverter.swift [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+import Foundation
+
+/// `RSDUnitConverter` is a static struct for defined unit converters for converting between US English (imperial)
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDUnitConverter.swift:36]: imperial (also on line below)

File: ResearchSuite/ResearchSuite/RSDUnitConverter.swift [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+/// storing the value in standardized metric units.
+public struct RSDUnitConverter {
+    
+    /// The converter to use for mass measurements.
+    public static let poundAndOunces = ImperialConverter<UnitMass>(largeUnit: .pounds, smallUnit: .ounces, baseUnit: .kilograms)
+    
+    /// The converter to use for length measurements.
+    public static let feetAndInches = ImperialConverter<UnitLength>(largeUnit: .feet, smallUnit: .inches, baseUnit: .centimeters)
+    
+    /// `ImperialConverter` is generic struct for converting measurements from various units.
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDUnitConverter.swift]: imperial

File: ResearchSuite/ResearchSuite/RSDUnitConverter.swift [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+    ///
+    /// -note: Imperial measurements are typically shown using a "larger" and "smaller" unit
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDUnitConverter.swift]: imperial

File: ResearchSuite/ResearchSuite/RSDUnitConverter.swift [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+    /// such as "ft, in" or "lb, oz". This converter will convert from a base unit (default
+    /// will be metric) to two values, one representing the larger unit and one representing
+    /// the smaller unit.
+    public struct ImperialConverter<UnitType> where UnitType : Dimension {
+        
+        /// The larger imperial unit. For example, "lb" (UnitMass) or "ft" (UnitLength).
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDUnitConverter.swift]: imperial

File: ResearchSuite/ResearchSuite/RSDUnitConverter.swift [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+        public let largeUnit: UnitType
+        
+        /// The smaller imperial unit. For example, "oz" (UnitMass) or "in" (UnitLength).
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDUnitConverter.swift:58]: imperial

File: ResearchSuite/ResearchSuite/RSDUnitConverter.swift [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+        /// such as height or weight.
+        public var baseUnit: UnitType
+        
+        /// The number of smaller units in the larger unit. For example, "1 lb = 12 oz" (UnitMass) or
+        /// "1 ft = 12 in" (UnitLength).
+        public let smallToLargeConversion: Int
+        
+        /// Default initializer.
+        /// - parameters:
+        ///     - largeUnit: The larger imperial unit.
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDUnitConverter.swift:71]: imperial (and next line)

File: ResearchSuite/ResearchSuite/RSDUnitConverter.swift [hunk 25]
```diff
@@ -0,0 +1,134 @@
...
+            if let ret = value as? Measurement<UnitType> {
+                return ret.converted(to: baseUnit)
+            } else if let num = (value as? NSNumber) ?? (value as? RSDJSONNumber)?.jsonNumber() {
+                return Measurement(value: num.doubleValue, unit: baseUnit)
+            } else {
+                return nil
+            }
+        }
+        
+        /// Convert the input Imperial values to a `Measurement` of the same unit type. This will convert
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDUnitConverter.swift]: Imperial

File: ResearchSuite/ResearchSuite/RSDUnitConverter.swift [hunk 26]
```diff
@@ -0,0 +1,134 @@
...
+            return (smallMeasurement + largeMeasurement.converted(to: smallUnit)).converted(to: baseUnit)
+        }
+        
+        /// Convert the input value to a tuple with the large and small measurement components used to
+        /// represent the measurement using Imperial units.
+        ///
+        /// - example:
+        /// ```
+        ///     // returns 1 foot, 8 inches
+        ///     let feet_1_AndInches_8 = lengthConverter.toImperialValue(from: 20 * 2.54)
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDUnitConverter.swift]: imperial

File: ResearchSuite/ResearchSuiteTests/Mock Objects/NSLocale+UnitTest.m [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+
+#import "NSLocale+UnitTest.h"
+#import <objc/runtime.h>
+
+static NSLocale * _currentTestLocale;
+static BOOL _hasBeenSwizzled = false;
+
+@implementation NSLocale (UnitTest)
+
++ (void)swizzleLocale {
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuiteTests/Mock Objects/NSLocale+UnitTest.m:42]: fun

**Review by Erin-Mounts — APPROVED** [review]
Nits. Most center around the use of  the term "Imperial" for non-Metric units. One comment contains an overly-detailed discussion of my pedantic concerns and why you might or might not want to consider different terminology, and most of the rest are just pointing out the places that would need editing if you decide you want to do that. (plus a couple of typos if I recall correctly, and a few comments/questions about the child vs adult units issues.) Otherwise looks good, merge when ready.

File: ResearchSuite/ResearchSuite/RSDTableItem.swift (hunk 18 above)
Comment (syoung-smallwisdom) [reply ResearchSuite/ResearchSuite/RSDTableItem.swift]: Our doctor records my 54" son's height using inches only. I thought about changing it, but decided that I would implement localization of formatted units instead since a question asked using this type is indicating a range such as, "Enter your child's height" and this needs to be something consistent. In my case, I go look at the report from his last checkup and write down what it says. Forcing the user to convert to 4' 6" is more confusing whereas having a text field input for this case where the text field formatter can convert from 4' 6" is easier for the user.

File: ResearchSuite/ResearchSuiteTests/Mock Objects/NSLocale+UnitTest.m (hunk 27 above)
Comment (syoung-smallwisdom) [reply ResearchSuite/ResearchSuiteTests/Mock Objects/NSLocale+UnitTest.m:42]: Sometimes ya just gotta go old-school.  :)
